### DISCUSSION TITLE
Why does the sdk create the Analytics instance right in the source code

### DISCUSSION THREAD

**Discussion by duannx** [rudderlabs/rudder-sdk-js#664] [Q&A]
The SDK creates an Analytics object [right in the source code](https://github.com/rudderlabs/rudder-sdk-js/blob/production/analytics.js#L1185) like that: 
```
const instance = new Analytics();
```
It leads to the code will be run immediately when the SDK is imported into the source code. I think it is an anti-pattern because the code runs without user awareness.

Example: running the SDK on a Vuejs SSR app. The SDK currently is not universal (Nodejs friendly) so it can not be run in the SSR context. The SSR will break whenever we import the SDK. The solution is to dynamic import the SDK on the client side. But it will lead to other problems like another separate chunk, difficulty when typing for the wrapper that dynamic imports the SDK...

**Comment by gitcommitshow** [comment]
I do not fully understand the suggestion here. Can you share the code sample for your SSR app and point out the issue there?

**Reply by duannx** [reply]
@gitcommitshow You just need to import the SDK in the Nodejs environment:
```
import RudderAnalytics from 'rudder-sdk-js'
export const AnalyticsWrapper = {
  track: RudderAnalytics.track,
}
```
The Nodejs will throw the error: `ReferenceError: window is not defined`. The reason as I said above is that the Rudderstack SDK automatically creates an Analytics object right in the source code. That runs sub-sequence code what uses the window object.

**Reply by gitcommitshow** [reply]
Got it. I guess, you already understand following but mentioning here for the clarity
1. `rudder-sdk-js` is a client-side library and built with the purpose of being used in browsers
2. For server-side tracking(sending events from the node backend), you should use [`rudder-sdk-node`](https://github.com/rudderlabs/rudder-sdk-node)
3. For server-side rendering, you can [follow this guide on GoogleAnalytics](https://stackoverflow.com/questions/60411351/how-to-use-google-analytics-with-next-js-app) OR [this guide by Gatsby on loading client-side only packages during SSR](https://www.gatsbyjs.com/docs/using-client-side-only-packages/) (although these guides are for react, but the principles remain the same)
4. We need to brainstorm on how to refactor the `rudder-sdk-js` code to make sure that it works in context of SSR without the need of any additional changes as per above point no. 3 (@duannx if you know how to do this, would you mind creating a PR?)

**Reply by duannx** [reply]
Thank you for understanding. 

I've already found a workaround. It is using the dynamic import on the client side only. But it has some tradeoffs as I mentioned above. So refactoring the JS SDK is still needed. I would love to work on it. But a PR now is too early. We should decide the way of refactoring first. 

My suggestion is to follow this rule strictly: `Definition only, don't execute anything`. As a library, we should provide the definition only, not the execution. And the developer who uses the library should have the right to decide where to execute it. The SDK should look like this: 
```js
// Just pure definition
export class Analytics {
   constructor(){
      ...
   }
   track(){}
   identify(){}
}
```
When using the package, the developer will decide when to execute it like this: 
```js
import Analytics from 'rudder-sdk-js'
if (isOnTheClientSide) {
   // The window object will be valid here
   const analytics = new Analytics()
   analytics.load(...)
}
```
The [segmentio/analytics-next](https://github.com/segmentio/analytics-next/blob/master/packages/browser/src/browser/index.ts) is a good example.

**Reply by gitcommitshow** [reply]
Looks like a good plan to me. Let's get more opinions on this.
@saikumarrs what do you think?

**Reply by saikumarrs** [reply]
With multiple work disruptions, I couldn't find enough time last week to look into this. I'll share my thoughts as soon as possible.

**Comment by saikumarrs** [comment]
Firstly, apologies for the delay.
@duannx Thanks for highlighting this. The suggestion you made makes sense and offers more flexibility to the SDK users. However, it might be a breaking change for the existing users and hence needs careful execution.

There is a plan in the pipeline for a significant code refactor and change the way SDK is served, but it may not happen very soon. We'll scope this out as part of that and address it accordingly.
CC @bardisg

Again, thanks for bringing this to our notice.

**Reply by duannx** [reply]
@saikumarrs I agreed. Could you please share the roadmap of `rudder-sdk-js`?

**Reply by saikumarrs** [reply]
@duannx, I've totally missed your response. Apologies for that.
Unfortunately, we do not yet have a public-facing roadmap for the JS SDK. However, I can assure you that we are aware of such issues and have bigger plans to execute in the coming quarters. We'll let everyone know once the time is right.
